### PR TITLE
allow style_selection() to work in RStudio's visual mode

### PR DESCRIPTION
**Pull request by kevinushey**
Closes https://github.com/r-lib/styler/issues/669.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -26,6 +26,7 @@
 - overhaul pgkdown site: Add search (#623), group function in Reference (#625).
 - always strip trailing spaces and make cache insensitive to it (#626).
 - typos in documentation (#643, #618, #614).
+- `style_selection()` now uses RStudio's new selection-based API when available
```
Comment (lorenzwalthert) [inline NEWS.md:29]:
```suggestion
- `style_selection()` now uses RStudio's new selection-based API when available (#670).
```

File: R/addins.R [hunk 2]
```diff
@@ -105,6 +112,31 @@ style_selection <- function() {
...
+  if (Sys.getenv("save_after_styling") == TRUE &&
+      !is.null(rstudioapi::documentPath(id))) {
+    invisible(rstudioapi::documentSave(id))
+  }
+
+}
+
+rstudio_has_selection_apis <- function() {
+
+  rstudioapi::hasFun("selectionGet") &&
```
Comment (lorenzwalthert) [inline R/addins.R:133]: I am afraid this is a bit slow (given that styler is already quite slow). Is making it RStudio version dependent an option?

File: R/addins.R [hunk 3]
```diff
@@ -91,6 +91,13 @@ style_active_pkg <- function() {
 #' @importFrom rlang abort
 #' @keywords internal
 style_selection <- function() {
+  if (rstudio_has_selection_apis())
+    style_selection_v2()
+  else
+    style_selection_v1()
+}
+
+style_selection_v1 <- function() {
```
Comment (lorenzwalthert) [inline R/addins.R:100]: Why v1 and v2? Does this refer to the `rstudioapi` major version?

File: R/addins.R [hunk 4]
```diff
@@ -105,6 +112,31 @@ style_selection <- function() {
...
 }
 
+style_selection_v2 <- function() {
+  communicate_addins_style_transformers()
+  id <- rstudioapi::documentId()
+  selection <- rstudioapi::selectionGet(id = id)
+  text <- selection$value
+  if (!nzchar(text)) abort("No code selected")
+  out <- style_text(text, transformers = get_addins_style_transformer())
+  rstudioapi::selectionSet(value = out, id = id)
```
Comment (lorenzwalthert) [inline R/addins.R:122]: What we did above with `if (context$selection[[1]]$range$end[2] == 1) "")` guarantees that when your selection is 
Screenshot 2020-09-18 at 23 44 40

There is exactly one line break after the selected area (and no blank line), and if it is 
Screenshot 2020-09-18 at 23 45 06
there is also a line break after the selection and no blank line. Does your implementation guarantee this as well?

File: R/addins.R [hunk 5]
```diff
@@ -105,6 +112,31 @@ style_selection <- function() {
...
+
+  if (Sys.getenv("save_after_styling") == TRUE &&
```
Comment (lorenzwalthert) [inline R/addins.R:124]: Note to self: when merged, update #631.

**Review by lorenzwalthert — CHANGES_REQUESTED** [review]
Thanks @kevinushey, I really appreciate RStudios proactive contribution. Below are a few questions. Also, these addin functions are currently mostly tested manually. If there are any ways to automate some of them, I'd be glad to hear about it (I only thought about mocking `get_rstudio_context()` so far), but nothing implemented. Maybe I can implement them in another PR so we can make sure your PR passes these tests. There are a few tricky cases with line breaks (last line blank, not blank, screenshots below)and selections that we need to get right. There are some instructions in `tests/testmanual/tests` you could try to run for now.

File: R/addins.R (hunk 2 above)
Comment (kevinushey) [reply R/addins.R:133]: Yes, that would also be suitable (`rstudioapi >= 0.12`; `RStudio >= 1.4.845`). Or, this condition could be cached and reused.

File: R/addins.R (hunk 3 above)
Comment (kevinushey) [reply R/addins.R:100]: Just arbitrary versions to indicate one is "newer" and one is "older".

File: R/addins.R (hunk 4 above)
Comment (kevinushey) [reply R/addins.R:122]: It doesn't; the new API doesn't provide any mechanism for expanding on modifying the user's current selection currently. We should probably discuss this on https://github.com/rstudio/rstudio/issues/7465.

**Comment by kevinushey** [comment]
We've made some changes to the development version of RStudio that should allow `styler` to continue working as-is, without any changes required from your side. Sorry for the trouble!

**Comment by lorenzwalthert** [comment]
Ok great. No worries.

**Pull request closed without merging**